### PR TITLE
MIR: split Operand::Consume into Copy and Move.

### PR DESCRIPTION
**Pull request by eddyb**
By encoding the choice of leaving the source untouched (`Copy`) and invalidating it (`Move`) in MIR, we can express moves of copyable values and have MIR borrow-checking enforce them, *including* ownership transfer of stack locals in calls  (when the ABI passes by indirection).

Optimizations could turn a "last-use" `Copy` into a `Move`, and the MIR borrow-checker, at least within the confines of safe code, could even do this when the underlying lvalue was borrowed.
(However, that last part would be the first time lifetime inference affects code generation, AFAIK).

Furthermore, as `Move`s invalidate borrows as well, for any local that is initialized only once, we can ignore borrows that happened before a `Move` and safely reuse/replace its memory storage.
This will allow us to perform NRVO in the presence of short-lived borrows, unlike LLVM (currently), and even compute optimal `StorageLive...StorageDead` ranges instead of discarding them.

### CODE REVIEW

File: src/librustc/mir/mod.rs [hunk 1]
```diff
@@ -1274,7 +1274,8 @@ pub struct VisibilityScopeData {
 /// being nested in one another.
 #[derive(Clone, PartialEq, RustcEncodable, RustcDecodable)]
 pub enum Operand<'tcx> {
-    Consume(Lvalue<'tcx>),
+    Copy(Lvalue<'tcx>),
```
Comment (nikomatsakis) [inline src/librustc/mir/mod.rs]: Can we leave some comments explaining what these mean, in terms of the expectations they provide? I think something like this:

- Copy: The value must be available for use afterwards. This implies that the type of the lvalue must be `Copy`; this is true by construction during build, but checked by the MIR type checker.
- Move: The value will not be used again. Safe for values of all types (modulo future developments towards `?Move`). Enforced by borrow checker. Sometimes Copy is converted to Move to enable "last-use" optimizations.

File: src/librustc_mir/borrow_check.rs [hunk 2]
```diff
@@ -1413,22 +1375,22 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+                                match *lv {
+                                    Operand::Copy(Lvalue::Local(l)) |
+                                    Operand::Move(Lvalue::Local(l)) if local == l => {
+                                        debug!(
+                                            "find_closure_span: found captured local {:?}",
+                                            l
+                                        );
+                                        return Some(v.span);
                                     }
+                                    _ => {}
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check.rs]: Personally, I'd rather this exhaustive (and perhaps move the `if` on the arm above inside the arm. But I rate the probability of overlooking this `match` after some future change to `Operand` as fairly low, so it's not that important.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
r=me modulo nits

**Comment by scottmcm** [comment]
FYI: If PR #46093 goes in first, I think it will conflict with this in a way that git won't notice.  This line will need to become a Move: https://github.com/rust-lang/rust/pull/46093/files#diff-69bf2a484158ebf56322b7ffb03601cdR108

**Review by nikomatsakis — APPROVED** [review]

**Comment by scottmcm** [comment]
Looks like this might be real?
```rust
[00:19:46] error[E0599]: no associated item named `Consume` found for type `rustc::mir::Operand<'_>` in the current scope
[00:19:46]    --> /checkout/src/librustc_mir/transform/add_moves_for_packed_drops.rs:135:34
[00:19:46]     |
[00:19:46] 135 |                      Rvalue::Use(Operand::Consume(location.clone())));
[00:19:46]     |                                  ^^^^^^^^^^^^^^^^
```

**Pull request merged**